keps/sig-api-machinery/3157-watch-list/README.md

```diff
@@ -839,7 +851,11 @@ conversion tests if API types are being modified.
 This section must be completed when targeting beta to a release.
 -->
 ###### How can a rollout or rollback fail? Can it impact already running workloads?
+Feature does not have a direct impact on rollout/rollback. 
+However, the faulty behavior of a feature can lead to the incorrect functioning of kube-controller-manager. 
+The main issues can arise during the initial informer synchronization, which may result in controller failures.
 
+Furthermore, if data consistency issues arise, such as missing data, the controllers simply do not consider the missing data.
```

Comment:
I don't understand what do you mean here - can you clarify?

Reply:
Do we have a way to detect this condition?  "weird" results are difficult to react to.  Should we periodically list and watchlist a relatively low cardinality resource and compare?  Perhaps Leases?  Given a specified RV, the result should be identical, right?

Reply:
Having this kind of second-line-of-defense would make sense in tests - I'm not sure we want to enable that in production code...

Reply:
Why can't we rely on the tests at all levels i.e. e2e, conformance, perf? After enabling this feature for `KCM`, if we don't find regressions, we should be fine, right?

Also the test described by David is already implemented (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apimachinery/watchlist.go#L82)

Reply:
> How will we be confident in the sameness in our production clusters as we make decisions about graduation? This feature is unique in its ability to compared against an existing feature in production.

Having ability to compare would be useful to enable in all our tests.
But I'm not keen on enabling it on real production clusters as this would actually potentially even double the load on control-plane and has a huge potential to do more harm than good.

Reply:
> But I'm not keen on enabling it on real production clusters as this would actually potentially even double the load on control-plane

I'm talking about a single, relatively low cardinality list on a small resource.  Doubling the load sounds like an extreme estimate.

> Why can't we rely on the tests at all levels i.e. e2e, conformance, perf?

The watch cache has disabused me of the notion that all the other tests are sufficient.

Reply:
We have a mutation informer check we turn on in CI.   Could a similar thing be built?

Reply:
We already have a "consistency-check" e2e test, ptal, https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apimachinery/watchlist.go#L82

Reply:
> I'm talking about a single, relatively low cardinality list on a small resource. Doubling the load sounds like an extreme estimate.

I'm doubtful in finding issues if we do that only for low-cardinality resource on small resources - the corner cases will usually happen on larger stuff.

> We have a mutation informer check we turn on in CI. Could a similar thing be built?

That's what I was trying to suggest - it makes sense to do that even for all resources in CI (or in clusters that I opt-in explicitly). I just don't want to do that for every production cluster.

Reply:
From my perspective, this is the only part that I want clarify on.
[But David is the approver and he may have other requests too]

Reply:
He should be online soon, and I'm okay with adding the mechanism you described to the beta criteria. I think David will like it as well.

Reply:
Let's make a check beta criteria.  We can do it like we do for the KUBE_CACHE_MUTATION_DETECTOR and enable it in CI as we do for that one.

Reply:
ok, added the check to the beta criteria.